File: py-algorithm/hashtable/single_number.py
import unittest

from math import gcd
from fractions import Fraction
from typing import List
from queue import PriorityQueue
from collections import Counter

"""
    https://leetcode.com/problems/single-number/
    
    Tìm số duy nhất trong mảng
"""


class Solution(unittest.TestCase):

    def singleNumber_OK(self, nums: List[int]) -> int:
        counter = Counter(nums)
        for key, value in counter.items():
            if value == 1: return key

        return -1

    # Cách này chỉ đúng nếu các số khác xuất hiện đúng 2 lần, số cần tìm xuất hiện 1 lần.
    def singleNumber(self, nums: List[int]) -> int:
        result = nums[0]
        for i in range(1, len(nums)):
            result ^= nums[i]

        return result

    def test(self):
        self.assertEqual(2, self.singleNumber([1, 2, 1]))


if __name__ == '__main__':
    unittest.main()

File: leetcode/src/main/java/leetcode/map/SingleNumber.java
package leetcode.map;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * https://leetcode.com/problems/single-number/
 * <p>
 * Tìm số xuất hiện duy nhất trong mảng
 */
public class SingleNumber {

    public int singleNumber_OK(int[] nums) {
        Map<Integer, Integer> counter = new HashMap<>();
        for (int num : nums) {
            counter.put(num, counter.getOrDefault(num, 0) + 1);
        }
        for (int num : counter.keySet()) {
            if (counter.get(num) == 1) return num;
        }
        return -1;
    }

    // If a = b -> a^b = 1
    public int singleNumber(int[] nums) {
        int result = nums[0];
        for (int i = 1; i < nums.length; i++) {
            result ^= nums[i];
        }
        return result;
    }

    @Test
    public void test() {
        assertEquals(2, singleNumber(new int[]{1, 1, 2}));
    }

}
